Automattic/jetpack
# Admin Menu: Do not alter menu slugs

Fixes https://github.com/Automattic/wp-calypso/issues/50950
Fixes https://github.com/Automattic/wp-calypso/issues/50945
Fixes https://github.com/Automattic/wp-calypso/issues/51018
Fixes https://github.com/Automattic/wp-calypso/issues/51024
Fixes https://github.com/Automattic/wp-calypso/issues/50939
Fixes https://github.com/Automattic/wp-calypso/issues/49915

### Changes proposed in this Pull Request:

Turns out that changing the menu slugs introduces a lot of side effects, mainly because the page hooks are still attached to the previous slugs:
- https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/plugin.php#L1469
- https://core.trac.wordpress.org/browser/trunk/src/wp-admin/menu-header.php#L245
- https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/plugin.php#L1331

To work around that, this PR explores a simpler approach: do not alter the menu slugs. Instead, we simply change the URL of the first submenu, since [that's the URL that's used in the parent menu](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/menu-header.php?rev=49193#L152). That removed the necessity of migrating submenus to new slugs, allowing us to have much less code to maintain which hopefully will make Nav Unification more robust.

### Jetpack product discussion
N/A.

### Does this pull request change what data or activity we track or use?
No.

### Testing instructions:

**Given the amount of changes, please take some time to smoke test on different sites and make sure there are no regressions on Calypso and WP Admin with [the "Show advanced dashboard pages" toggle on and off](https://wordpress.com/me/account).**

_Simple_
- Apply D58455-code to your WP.com sandbox.
- Sandbox the API and a simple site.
- Go to https://wordpress.com and switch to the sandboxed site.
- Check the menu under different scenarios (Calypso, WP Admin, toggle on/off) and make sure there are no regressions.

_Atomic_
- Install Jetpack Beta on a Atomic site and switch to the branch of this PR.
- Go to https://wordpress.com and switch to the WoA site.
- Check the menu under different scenarios (Calypso, WP Admin, toggle on/off) and make sure there are no regressions.

_Jetpack_
- Spin up a Jetpack site running this branch.
- Spin up a local instance of Calypso with [this code](https://github.com/Automattic/wp-calypso/blob/46541a684523325e4efd293d0f8b7f2530f4e6f2/client/state/selectors/is-nav-unification-enabled.js#L27-L31) removed/commented out.
- Go to http://calypso.localhost:3000 and switch to the Jetpack site.
- Check the menu in Calypso only (since Nav Unification is disabled on WP Admin) and make sure there are no regressions (all menus should point to Calypso regardless of the toggle, except Feedback, Customize, Import, Export and WP Admin).

After confirming there are no regressions, follow the instructions below to confirm that this indeed fixes the linked issues.

**https://github.com/Automattic/wp-calypso/issues/50950, https://github.com/Automattic/wp-calypso/issues/51018, https://github.com/Automattic/wp-calypso/issues/51024:**

- On the Atomic site, go to `/wp-admin/options-general.php`.
- Change the site language to a non-English language such as Spanish or Italian.
- Install the following plugins/themes:
  - [WP Headers and Footers](https://wordpress.org/plugins/wp-headers-and-footers/).
  - [UpdraftPlus](https://wordpress.org/plugins/updraftplus/).
  - [Compress JPEG & PNG images](https://wordpress.org/plugins/tiny-compress-images/).
  - [Commenter Emails](https://wordpress.org/plugins/commenter-emails/).
  - [Storefront](https://woocommerce.com/storefront/).
- Go to Jetpack > Settings > Sharing, and enable either "Publicize connections" or "Sharing buttons".
- Inspect the URL of the "Settings > WP Headers and Footers" submenu and make sure it points to `/wp-admin/options-general.php?page=wp-header-and-footer`.
- Click on it and make sure the page loads successfully.
- Inspect the URL of the "Settings > UpdraftPlus" submenu and make sure it points to `/wp-admin/options-general.php?page=updraftplus`.
- Click on it and make sure the page loads successfully.
- Inspect the URL of the "Settings > Sharing" submenu and make sure it points to `/wp-admin/options-general.php?page=sharing`.
- Click on it and make sure the page loads successfully.
- Click on "Media" (the parent menu, not the first submenu).
- Make sure the media library loads successfully.
- Click on "Comments" (the parent menu, not the first submenu).
- Make sure the comments page loads successfully.
- Inspect the URL of the "Appearance > Storefront" submenu and make sure it points to `/wp-admin/themes.php?page=storefront-welcome`.
- Click on it and make sure the page loads successfully.

**https://github.com/Automattic/wp-calypso/issues/50945**
- Make sure My Home is always visible on Simple and Atomic sites, even when the "Show advanced dashboard pages" toggle is off.
- Make sure there is no link to WP Admin on Simple and Atomic sites, but it's still accessible via the `/wp-admin` URL.

**https://github.com/Automattic/wp-calypso/issues/50939, https://github.com/Automattic/wp-calypso/issues/49915:**
- Install the Ovation theme on an Atomic site.
- Make sure there are no duplicated items under the Appearance menu under different scenarios (Calypso, WP Admin, toggle on/off).
